beacon-chain/sync/data_columns_sampling.go

```diff
@@ -4,231 +4,254 @@ import (
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/startup"
 	fieldparams "github.com/prysmaticlabs/prysm/v5/config/fieldparams"
 	"github.com/prysmaticlabs/prysm/v5/config/params"
 	"github.com/prysmaticlabs/prysm/v5/crypto/rand"
 	eth "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
 	"github.com/prysmaticlabs/prysm/v5/runtime/version"
 )
 
+const PeerRefreshInterval = 1 * time.Minute
+
 type roundSummary struct {
 	RequestedColumns []uint64
 	MissingColumns   map[uint64]bool
 }
 
-// randomizeColumns returns a slice containing all columns in a random order.
-func randomizeColumns(columns map[uint64]bool) []uint64 {
-	// Create a slice from columns.
-	randomized := make([]uint64, 0, len(columns))
-	for column := range columns {
-		randomized = append(randomized, column)
-	}
-
-	// Shuffle the slice.
-	rand.NewGenerator().Shuffle(len(randomized), func(i, j int) {
-		randomized[i], randomized[j] = randomized[j], randomized[i]
-	})
-
-	return randomized
+// DataColumnSampler defines the interface for sampling data columns from peers for requested block root and samples count.
+type DataColumnSampler interface {
+	// Run starts the data column sampling service.
+	Run(ctx context.Context)
 }
 
-// sortedSliceFromMap returns a sorted slices of keys from a map.
-func sortedSliceFromMap(m map[uint64]bool) []uint64 {
-	result := make([]uint64, 0, len(m))
-	for k := range m {
-		result = append(result, k)
-	}
+var _ DataColumnSampler = (*dataColumnSampler1D)(nil)
 
-	sort.Slice(result, func(i, j int) bool {
-		return result[i] < result[j]
-	})
+// dataColumnSampler1D implements the DataColumnSampler interface for PeerDAS 1D.
+type dataColumnSampler1D struct {
+	sync.RWMutex
 
-	return result
-}
+	p2p           p2p.P2P
+	clock         *startup.Clock
+	ctxMap        ContextByteVersions
+	stateNotifier statefeed.Notifier
 
-// custodyColumnsFromPeer returns the columns the peer should custody.
-func (s *Service) custodyColumnsFromPeer(pid peer.ID) (map[uint64]bool, error) {
-	// Retrieve the custody count of the peer.
-	custodySubnetCount := s.cfg.p2p.CustodyCountFromRemotePeer(pid)
+	// custodyColumns is a set of columns that the node is responsible for custody.
+	custodyColumns map[uint64]bool
+	// columnFromPeer maps a peer to the columns it is responsible for custody.
+	columnFromPeer map[peer.ID]map[uint64]bool
+	// peerFromColumn maps a column to the peer responsible for custody.
+	peerFromColumn map[uint64]map[peer.ID]bool
+}
 
-	// Extract the node ID from the peer ID.
-	nodeID, err := p2p.ConvertPeerIDToNodeID(pid)
-	if err != nil {
-		return nil, errors.Wrap(err, "extract node ID")
+// newDataColumnSampler1D creates a new 1D data column sampler.
+func newDataColumnSampler1D(
+	p2p p2p.P2P,
+	clock *startup.Clock,
+	ctxMap ContextByteVersions,
+	stateNotifier statefeed.Notifier,
+) *dataColumnSampler1D {
+	columnToPeerMap := make(map[uint64]map[peer.ID]bool, params.BeaconConfig().NumberOfColumns)
```

Comment:
`params.BeaconConfig().NumberOfColumns` can be extracted in a variable as it is used multiple times in the function.